File: server/internal/accounts/simplefin_types.go
package accounts

import (
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"time"

	"tallyo/internal/clients"
	"tallyo/internal/graph/model"
	u "tallyo/internal/utils"

	"github.com/samber/lo"
)

// typeKeywords maps account name keywords to an AccountType. Order matters:
// more specific patterns should come before more generic ones.
var (
	depositoryRE        = regexp.MustCompile(`(?i)\b(checking|savings|money\s+market|cd|hsa|cash\s+mgmt|cash\s+management|high\s+yield)\b`)
	creditRE            = regexp.MustCompile(`(?i)\b(card|credit|visa|mastercard|amex|american\s+express|freedom|sapphire|platinum|gold|blue\s+cash|hyatt|gateway|mileage|rewards|cashback)\b`)
	loanRE              = regexp.MustCompile(`(?i)\b(loan|mortgage|auto|heloc|student|line\s+of\s+credit)\b`)
	simpleFinNameMaskRE = regexp.MustCompile(`\s+\(([0-9]+)\)\s*$`)
)

// SimpleFinAccessTokenSecret carries the access_url and scheduling state for
// use by the sync adapter; it is never exposed via GraphQL.
type SimpleFinAccessTokenSecret struct {
	ID                int64
	AccessURL         string
	OwnerID           int64
	SyncCron          string
	LastSyncedAt      *time.Time
	NextSyncAt        *time.Time
	NextBalanceSyncAt *time.Time
}

// SimpleFinAccountFields is the inferred account data for a SimpleFIN account.
type SimpleFinAccountFields struct {
	ID          string
	Name        string
	Type        model.AccountType
	Mask        *string
	NeedsReview bool
}

// FieldsFromSimpleFinAccount infers account fields from a SimpleFIN account.
// Type is inferred from holdings, the account name, and the balance sign.
func FieldsFromSimpleFinAccount(account clients.SimpleFinAccount) SimpleFinAccountFields {
	balance, _ := strconv.ParseFloat(strings.TrimSpace(account.Balance), 64)
	name, mask := simpleFinNameAndMask(account.Name)
	accountType, needsReview := inferSimpleFinAccountType(name, account.Holdings, balance)
	return SimpleFinAccountFields{
		ID:          account.ID,
		Name:        name,
		Type:        accountType,
		Mask:        mask,
		NeedsReview: needsReview,
	}
}

func simpleFinNameAndMask(name string) (string, *string) {
	matches := simpleFinNameMaskRE.FindStringSubmatchIndex(name)
	if len(matches) == 0 {
		return strings.TrimSpace(name), nil
	}

	base := strings.TrimSpace(name[:matches[0]])
	if base == "" {
		return strings.TrimSpace(name), nil
	}

	mask := name[matches[2]:matches[3]]
	return base, &mask
}

// inferSimpleFinAccountType classifies a SimpleFIN account once, at creation.
// Holdings mark an investment account. Otherwise name keywords win (the most
// specific signal, and the only way to distinguish LOAN from CREDIT or to keep
// an overdrawn checking account as DEPOSITORY). When the name is uninformative
// we fall back to the balance sign: SimpleFIN reports liabilities as negative
// and assets as positive. A zero balance with no name signal is an unknown we
// default to CREDIT and flag for manual review.
func inferSimpleFinAccountType(
	name string,
	holdings []clients.SimpleFinHolding,
	balance float64,
) (model.AccountType, bool) {
	if len(holdings) > 0 {
		return model.AccountTypeInvestment, false
	}
	lower := strings.ToLower(name)
	switch {
	case strings.Contains(lower, "brokerage"):
		return model.AccountTypeInvestment, false
	case depositoryRE.MatchString(lower):
		return model.AccountTypeDepository, false
	case creditRE.MatchString(lower):
		return model.AccountTypeCredit, false
	case loanRE.MatchString(lower):
		return model.AccountTypeLoan, false
	}
	switch {
	case balance < 0:
		return model.AccountTypeCredit, false
	case balance > 0:
		return model.AccountTypeDepository, false
	default:
		return model.AccountTypeCredit, true
	}
}

func simpleFinOrgDomain(rawURL string) string {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return rawURL
	}
	return parsed.Hostname()
}

func NewUpsertSimpleFinConnectionParams(tokenID int64, ownerID int64, conn clients.SimpleFinConnection) UpsertSimpleFinConnectionParams {
	return UpsertSimpleFinConnectionParams{
		ExternalID:    conn.ConnID,
		AccessTokenID: tokenID,
		OrgID:         lo.EmptyableToPtr(conn.OrgID),
		OrgDomain:     lo.EmptyableToPtr(simpleFinOrgDomain(conn.OrgURL)),
		OrgURL:        lo.EmptyableToPtr(conn.OrgURL),
		SfinURL:       lo.EmptyableToPtr(conn.SfinURL),
		LogoURL:       u.DuckDuckGoFaviconURL(conn.OrgURL),
		Name:          conn.Name,
		OwnerID:       ownerID,
	}
}

File: server-rs/src/accounts/simplefin_types.rs
use url::Url;

use crate::{
    accounts::AccountType,
    clients::simplefin::{SimpleFinAccount, SimpleFinConnection},
    database::queries,
    utils::favicon::duckduckgo_favicon_url,
};

pub use crate::database::queries::SimpleFinTokenSecretsRow as SimpleFinAccessTokenSecret;

impl From<queries::SimpleFinTokenSecretByConnIdRow> for SimpleFinAccessTokenSecret {
    fn from(row: queries::SimpleFinTokenSecretByConnIdRow) -> Self {
        Self {
            access_url: row.access_url,
            created_at: row.created_at,
            id: row.id,
            label: row.label,
            last_synced_at: row.last_synced_at,
            next_balance_sync_at: row.next_balance_sync_at,
            next_sync_at: row.next_sync_at,
            owner_id: row.owner_id,
            sync_cron: row.sync_cron,
            updated_at: row.updated_at,
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct SimpleFinAccountFields {
    pub id: String,
    pub name: String,
    pub account_type: AccountType,
    pub mask: Option<String>,
    pub needs_review: bool,
}

#[derive(Clone, Debug, PartialEq)]
pub struct UpsertSimpleFinConnectionParams {
    pub external_id: String,
    pub access_token_id: i64,
    pub org_id: Option<String>,
    pub org_domain: Option<String>,
    pub org_url: Option<String>,
    pub sfin_url: Option<String>,
    pub logo_url: Option<String>,
    pub name: String,
    pub owner_id: i64,
}

pub fn fields_from_simplefin_account(account: &SimpleFinAccount) -> SimpleFinAccountFields {
    let (name, mask) = simplefin_name_and_mask(&account.name);
    let balance = account.balance.trim().parse::<f64>().unwrap_or_default();
    let (account_type, needs_review) = infer_simplefin_account_type(&name, !account.holdings.is_empty(), balance);
    SimpleFinAccountFields {
        id: account.id.clone(),
        name,
        account_type,
        mask,
        needs_review,
    }
}

pub fn new_upsert_simple_fin_connection_params(
    token_id: i64,
    owner_id: i64,
    connection: &SimpleFinConnection,
) -> UpsertSimpleFinConnectionParams {
    let org_domain = simplefin_org_domain(&connection.org_url);
    UpsertSimpleFinConnectionParams {
        external_id: connection.conn_id.clone(),
        access_token_id: token_id,
        org_id: nonempty(&connection.org_id),
        org_domain: nonempty(&org_domain),
        org_url: nonempty(&connection.org_url),
        sfin_url: nonempty(&connection.sfin_url),
        logo_url: duckduckgo_favicon_url(&connection.org_url),
        name: connection.name.clone(),
        owner_id,
    }
}

fn simplefin_name_and_mask(name: &str) -> (String, Option<String>) {
    let trimmed = name.trim_end();
    let Some(without_closing) = trimmed.strip_suffix(')') else {
        return (name.trim().to_owned(), None);
    };
    let Some((base, mask)) = without_closing.rsplit_once('(') else {
        return (name.trim().to_owned(), None);
    };
    if !base.chars().next_back().is_some_and(char::is_whitespace)
        || mask.is_empty()
        || !mask.bytes().all(|byte| byte.is_ascii_digit())
    {
        return (name.trim().to_owned(), None);
    }
    let base = base.trim();
    if base.is_empty() {
        return (name.trim().to_owned(), None);
    }
    (base.to_owned(), Some(mask.to_owned()))
}

fn infer_simplefin_account_type(name: &str, has_holdings: bool, balance: f64) -> (AccountType, bool) {
    let name = name
        .to_ascii_lowercase()
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ");
    match (has_holdings, name.as_str(), balance) {
        (true, _, _) => (AccountType::Investment, false),
        (_, name, _) if name.contains("brokerage") => (AccountType::Investment, false),
        (_, name, _)
            if has_keyword(
                name,
                &[
                    "checking",
                    "savings",
                    "money market",
                    "cd",
                    "hsa",
                    "cash mgmt",
                    "cash management",
                    "high yield",
                ],
            ) =>
        {
            (AccountType::Depository, false)
        }
        (_, name, _)
            if has_keyword(
                name,
                &[
                    "card",
                    "credit",
                    "visa",
                    "mastercard",
                    "amex",
                    "american express",
                    "freedom",
                    "sapphire",
                    "platinum",
                    "gold",
                    "blue cash",
                    "hyatt",
                    "gateway",
                    "mileage",
                    "rewards",
                    "cashback",
                ],
            ) =>
        {
            (AccountType::Credit, false)
        }
        (_, name, _)
            if has_keyword(
                name,
                &["loan", "mortgage", "auto", "heloc", "student", "line of credit"],
            ) =>
        {
            (AccountType::Loan, false)
        }
        (_, _, balance) if balance < 0.0 => (AccountType::Credit, false),
        (_, _, balance) if balance > 0.0 => (AccountType::Depository, false),
        _ => (AccountType::Credit, true),
    }
}

fn has_keyword(name: &str, keywords: &[&str]) -> bool {
    keywords.iter().any(|keyword| {
        name.match_indices(keyword).any(|(index, _)| {
            let before = name[..index].chars().next_back();
            let after = name[index + keyword.len()..].chars().next();
            !before.is_some_and(char::is_alphanumeric) && !after.is_some_and(char::is_alphanumeric)
        })
    })
}

fn simplefin_org_domain(url: &str) -> String {
    Url::parse(url)
        .ok()
        .and_then(|parsed| parsed.host_str().map(ToOwned::to_owned))
        .unwrap_or_else(|| url.to_owned())
}

fn nonempty(value: &str) -> Option<String> {
    (!value.is_empty()).then(|| value.to_owned())
}

#[cfg(test)]
mod tests {
    use crate::{
        accounts::{AccountType, simplefin_types::fields_from_simplefin_account},
        clients::simplefin::{SimpleFinAccount, SimpleFinHolding},
    };

    #[test]
    fn infers_simplefin_account_fields() {
        struct Case {
            name: &'static str,
            balance: &'static str,
            holdings: bool,
            expected_name: &'static str,
            expected_mask: Option<&'static str>,
            expected_type: AccountType,
            expected_needs_review: bool,
        }

        let cases = [
            Case {
                name: "Joint Checking",
                balance: "",
                holdings: false,
                expected_name: "Joint Checking",
                expected_mask: None,
                expected_type: AccountType::Depository,
                expected_needs_review: false,
            },
            Case {
                name: "Sapphire Reserve (4628)",
                balance: "",
                holdings: false,
                expected_name: "Sapphire Reserve",
                expected_mask: Some("4628"),
                expected_type: AccountType::Credit,
                expected_needs_review: false,
            },
            Case {
                name: "Sapphire Reserve",
                balance: "",
                holdings: false,
                expected_name: "Sapphire Reserve",
                expected_mask: None,
                expected_type: AccountType::Credit,
                expected_needs_review: false,
            },
            Case {
                name: "Auto Loan",
                balance: "",
                holdings: false,
                expected_name: "Auto Loan",
                expected_mask: None,
                expected_type: AccountType::Loan,
                expected_needs_review: false,
            },
            Case {
                name: "Self Directed",
                balance: "",
                holdings: true,
                expected_name: "Self Directed",
                expected_mask: None,
                expected_type: AccountType::Investment,
                expected_needs_review: false,
            },
            Case {
                name: "Taxable brokerage",
                balance: "",
                holdings: false,
                expected_name: "Taxable brokerage",
                expected_mask: None,
                expected_type: AccountType::Investment,
                expected_needs_review: false,
            },
            Case {
                name: "Self Directed BrokerageAccount",
                balance: "",
                holdings: false,
                expected_name: "Self Directed BrokerageAccount",
                expected_mask: None,
                expected_type: AccountType::Investment,
                expected_needs_review: false,
            },
            Case {
                name: "Mystery Account",
                balance: "",
                holdings: false,
                expected_name: "Mystery Account",
                expected_mask: None,
                expected_type: AccountType::Credit,
                expected_needs_review: true,
            },
            Case {
                name: "Joint Checking",
                balance: "-106.17",
                holdings: false,
                expected_name: "Joint Checking",
                expected_mask: None,
                expected_type: AccountType::Depository,
                expected_needs_review: false,
            },
            Case {
                name: "Premier Account",
                balance: "-50.00",
                holdings: false,
                expected_name: "Premier Account",
                expected_mask: None,
                expected_type: AccountType::Credit,
                expected_needs_review: false,
            },
            Case {
                name: "Premier Account",
                balance: "500.00",
                holdings: false,
                expected_name: "Premier Account",
                expected_mask: None,
                expected_type: AccountType::Depository,
                expected_needs_review: false,
            },
        ];

        for case in cases {
            let fields = fields_from_simplefin_account(&SimpleFinAccount {
                name: case.name.into(),
                balance: case.balance.into(),
                holdings: if case.holdings { vec![SimpleFinHolding::default()] } else { Vec::new() },
                ..Default::default()
            });
            assert_eq!(fields.name, case.expected_name, "{}", case.name);
            assert_eq!(fields.mask.as_deref(), case.expected_mask, "{}", case.name);
            assert_eq!(fields.account_type, case.expected_type, "{}", case.name);
            assert_eq!(fields.needs_review, case.expected_needs_review, "{}", case.name);
        }
    }
}
